crates/fuel-core/src/database/balances.rs

```diff
@@ -48,37 +59,101 @@ impl StorageMutate<ContractsAssets> for Database {
         key: &<ContractsAssets as Mappable>::Key,
         value: &<ContractsAssets as Mappable>::Value,
     ) -> Result<Option<<ContractsAssets as Mappable>::OwnedValue>, Self::Error> {
-        Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
-            .map_err(Into::into)
+        let prev = Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
+            .map_err(Into::into);
+
+        // Get latest metadata entry
+        let prev_metadata = self
```

Comment:
Why do we need `iter_all` if we have only one entry of the `SparseMerkleMetadata` in the database for the `ContractsAssetsMerkleMetadata` table?

I think you need to use `self.storage::<ContractsAssetsMerkleMetadata>().get(key.contract_id())`. Because now it seems you take the state of any last contract=)

Reply:
Yes, I think you're right! I will make that change.